pkg/cmd/codespace/ssh.go

```diff
@@ -147,6 +161,145 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
+
+		if cs.State != "Available" {
+			fmt.Fprintf(os.Stderr, "skipping unavailable codespace %s: %s\n", cs.Name, cs.State)
+			status = cmdutil.SilentError
+			continue
+		}
+
+		sshUser, ok := sshUsers[cs.Repository.FullName]
+		if !ok {
+			session, err := openSSHSession(ctx, a, cs, nil)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error connecting to codespace: %v\n", err)
+
+				// Move on to the next codespace. We don't want to bail here - just because we're not
+				// able to set up connectivity to one doesn't mean we shouldn't make a best effort to
+				// generate configs for the rest of them.
+				status = cmdutil.SilentError
+				continue
+			}
+			defer session.Close()
+
+			a.StartProgressIndicatorWithLabel(fmt.Sprintf("Fetching SSH Details for %s", cs.Name))
+			_, sshUser, err = session.StartSSHServer(ctx)
+			a.StopProgressIndicator()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error getting ssh server details: %v", err)
+				status = cmdutil.SilentError
+				continue // see above
+			}
+			sshUsers[cs.Repository.FullName] = sshUser
+		}
+
+		conf := codespaceSSHConfig{
+			Name:       cs.Name,
+			EscapedRef: strings.ReplaceAll(cs.GitStatus.Ref, "/", "-"),
+			SSHUser:    sshUser,
+			GHExec:     ghexec,
+		}
+		if err := t.Execute(a.io.Out, conf); err != nil {
+			return err
+		}
+	}
+
+	return status
+}
+
+// codespaceSSHConfig contains values needed to write an OpenSSH host
+// configuration for a single codespace. For example:
+//
+// Host {{Name}}.{{EscapedRef}
+// User {{SSHUser}
+// ProxyCommand {{GHExec}} cs ssh -c {{Name}} --stdio
+//
+// EscapedRef is included in the name to help distinguish between codespaces
+// when tab-completing ssh hostnames. '/' characters in EscapedRef are
+// flattened to '-' to prevent problems with tab completion or when the
+// hostname appears in ControlMaster socket paths.
+type codespaceSSHConfig struct {
+	Name       string // the codespace name, passed to `ssh -c`
+	EscapedRef string // the currently checked-out branch
+	SSHUser    string // the remote ssh username
+	GHExec     string // path used for invoking the current `gh` binary
+}
+
+func openSSHSession(ctx context.Context, a *App, codespace *api.Codespace, liveshareLogger *log.Logger) (*liveshare.Session, error) {
+	// TODO(josebalius): We can fetch the user in parallel to everything else
+	// we should convert this call and others to happen async
+	user, err := a.apiClient.GetUser(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting user: %w", err)
+	}
+
+	authkeys := make(chan error, 1)
+	go func() {
+		authkeys <- checkAuthorizedKeys(ctx, a.apiClient, user.Login)
+	}()
+
+	if liveshareLogger == nil {
+		liveshareLogger = noopLogger()
+	}
+
+	session, err := codespaces.ConnectToLiveshare(ctx, a, liveshareLogger, a.apiClient, codespace)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to codespace: %w", err)
+	}
+
+	if err := <-authkeys; err != nil {
+		return nil, err
```

Comment:
You need an explicit session.Close call here. (Previously it was done by a defer.)

Reply:
no longer necessary after refactoring :)